pkg/hub/controllers/cluster/reconciler.go

```diff
@@ -458,3 +486,47 @@ func (r *Reconciler) InjectClient(c client.Client) error {
 	r.Client = c
 	return nil
 }
+
+func (r *Reconciler) handleClusterDeletion(cluster *hubv1alpha1.Cluster, ctx context.Context, req reconcile.Request) (bool, reconcile.Result, error) {
+	log := logger.FromContext(ctx)
+
+	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// registering our finalizer.
+		if !controllerutil.ContainsFinalizer(cluster, clusterDeregisterFinalizer) {
+			controllerutil.AddFinalizer(cluster, clusterDeregisterFinalizer)
+			if err := r.Update(ctx, cluster); err != nil {
+				return true, reconcile.Result{}, err
+			}
+		}
+	} else {
+		// The object is being deleted
+		if controllerutil.ContainsFinalizer(cluster, clusterDeregisterFinalizer) && !cluster.Status.IsDeregisterInProgress && retryAttempts <= 3 {
+			// our finalizer is present, so lets handle any external dependency
+			if err := r.createDeregisterJob(ctx, cluster); err != nil {
+				// unable to deregister the worker operator, return with an error and raise event
+				log.Error(err, "unable to deregister the worker operator")
+				// increment count for retryAttempts
+				retryAttempts++
+				// resetting isDeregisterInProgress to false
+				statusUpdateErr := r.updateRegistrationStatusAndDeregisterInProgress(ctx, cluster, hubv1alpha1.RegistrationStatusDeregisterFailed, false)
+				if statusUpdateErr != nil {
+					// log.Error(statusUpdateErr, "unable to update registration status")
+					recordEventErr := r.EventRecorder.RecordEvent(ctx, &events.Event{
+						Object:            cluster,
+						Name:              ossEvents.EventDeregitrationJobFailed,
```

Comment:
globally please correct EventDeregitrationJobFailed to EventDeregistrationJobFailed